test/src/tests-batchUploader.ts

```diff
@@ -89,6 +89,612 @@ describe('batch uploader', () => {
+
+            // Batch 2 should contain three custom events
+            expect(batch2.events.length).to.equal(3);
+            expect(batch2.events[0].data.event_name).to.equal('Test Event 1');
+            expect(batch2.events[1].data.event_name).to.equal('Test Event 2');
+            expect(batch2.events[2].data.event_name).to.equal('Test Event 3');
+
+            // Log a third batch of events
+            window.mParticle.logEvent('Test Event 4');
+            window.mParticle.logEvent('Test Event 5');
+
+            // Force an upload of events
+            window.mParticle.upload();
+
+            // Force time to pass so that batch processing occurs
+            clock.tick(1000);
+
+            const batch3 = JSON.parse(window.fetchMock._calls[2][1].body);
+
+            // Batch 3 should contain two custom events
+            expect(batch3.events.length).to.equal(2);
+            expect(batch3.events[0].data.event_name).to.equal('Test Event 4');
+            expect(batch3.events[1].data.event_name).to.equal('Test Event 5');
+
+            done();
+        });
+
+        it('should keep batches in sequence for future retries when an HTTP 500 error occurs', done => {
+            // If batches cannot upload, they should be added back to the Batch Queue
+            // in the order they were created so they can be retransmitted.
+
+            window.fetchMock.post(urls.eventsV3, 500);
+
+            window.mParticle.config.flags = {
+                eventsV3: '100',
+                eventBatchingIntervalMillis: 1000,
+            };
+
+            window.mParticle._resetForTests(MPConfig);
+            window.mParticle.init(apiKey, window.mParticle.config);
+            // Generates Batch 1 with Session Start + AST
+
+            // Adds a custom event to Batch 1
+            window.mParticle.logEvent('Test Event');
+
+            // Force an upload of events
+            window.mParticle.upload();
+
+            expect(
+                window.fetchMock.called(),
+                'FetchMock should have been called'
+            ).to.equal(true);
+
+            const batch1 = JSON.parse(window.fetchMock._calls[0][1].body);
+
+            // Batch 1 should contain only session start, AST and a single event
+            // in this exact order
+            expect(batch1.events.length).to.equal(3);
+            expect(batch1.events[0].event_type).to.equal('session_start');
+            expect(batch1.events[1].event_type).to.equal(
+                'application_state_transition'
+            );
+            expect(batch1.events[2].data.event_name).to.equal('Test Event');
+
+            // // Force time to pass so that batch processing occurs
+            // clock.tick(1000);
+
+            // Batch 2
+            window.mParticle.logEvent('Test Event 1');
+            window.mParticle.logEvent('Test Event 2');
+            window.mParticle.logEvent('Test Event 3');
+
+            // Force an upload of events
```

Comment:
I'd change this comment here and elsewhere to be a bit more specific because the upload process is very nuanced. 
 Something like the below:

```suggestion
            // Manually initiate the upload process - turn event into batches and upload the batch
```